File: 51-n-queens/n-queens.py
class Solution(object):
    def isSafe(self, board, row, col, n):
        # horizontal
        for j in range(n):
            if board[row][j] == 'Q':
                return False

        # vertical
        for i in range(n):
            if board[i][col] == 'Q':
                return False

        # left diagonal
        for i, j in zip(range(row - 1, -1, -1), range(col - 1, -1, -1)):
            if board[i][j] == 'Q':
                return False

        # right diagonal
        for i, j in zip(range(row - 1, -1, -1), range(col + 1, n)):
            if board[i][j] == 'Q':
                return False

        return True

    def nQueen(self, board, row, n, result):
        if row == n:
            result.append(["".join(row) for row in board])  # Add valid solution
            return

        for j in range(n):
            if self.isSafe(board, row, j, n):
                board[row][j] = 'Q'  # Place the queen

                self.nQueen(board, row + 1, n, result)  # Recur to place the next queen

                # Backtrack
                board[row][j] = '.'  # Reset the position

    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        result = []
        board = [['.' for _ in range(n)] for _ in range(n)]  # Initialize board with empty rows

        self.nQueen(board, 0, n, result)

        return result

File: 51-n-queens/n-queens.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    bool isSafe(vector<string>& board, int row, int col, int n) {
        // horizontal
        for (int j = 0; j < n; j++) {
            if (board[row][j] == 'Q') {
                return false;
            }
        }

        // vertical
        for (int i = 0; i < n; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }

        // left diagonal
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        // right diagonal
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        return true;
    }

    void nQueen(vector<string>& board, int row, int n, vector<vector<string>>& result) {
        if (row == n) {
            result.push_back(board);  // Found a valid solution
            return;
        }

        for (int j = 0; j < n; j++) {
            if (isSafe(board, row, j, n)) {
                board[row][j] = 'Q';  // Place the queen

                nQueen(board, row + 1, n, result);  // Recur to place the next queen

                // Backtrack
                board[row][j] = '.';  // Reset the position
            }
        }
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> result;
        vector<string> board(n, string(n, '.'));  // Initialize board with empty rows

        nQueen(board, 0, n, result);

        return result;
    }
};
